### DISCUSSION TITLE
MecrispCube for TinyGo

### DISCUSSION THREAD

**Discussion by sqvery** [tinygo-org/tinygo#4635] [Ideas]
I have found Mecrisp Forth extremely useful in modern embedded development.

When developing with a compiled language, the code-compile-upload-run cycle can be frustrating. This is where a modern Forth, like Mecrisp, shines. It’s fast enough and operates very close to the hardware, enabling you to toggle bits, read registers, and retry operations on the fly. However, it comes without many of the tools and libraries developers often rely on.

This is where [MecrispCube](https://www.spyr.ch/twiki/bin/view/MecrispCube) comes in. It is a compiled environment that incorporates Mecrisp assembly code, offering the best of both worlds.

### Questions

1. Does TinyGo include an RTOS or multitasking environment that would allow Forth to run as a [CMSIS-RTOS](https://www.spyr.ch/twiki/bin/view/MecrispCube/CmsisRtos) thread? Alternatively, would it be feasible to include TinyGo as a thread in a Forth-based cooperative multitasking system?

3. Is the register structure of TinyGo similar to that of C, enabling [calling C functions](https://www.spyr.ch/twiki/bin/view/MecrispCube/CallingCFunction) from Forth and vice versa?

5. Should I wait for TinyGo to mature?

### Simple Questions
6. How do I include GNU Assembler code in a TinyGo file?

7. Would it be simpler to use interrupts to trigger Mecrisp code execution?

8. Embedded systems often run in infinite loops. How would I call the Forth engine from within such a loop?

### General Thoughts
8. Am I completely barking up the wrong tree with this approach?

**Comment by aykevl** [comment]
Oh that's an interesting idea!

TinyGo has its own RTOS-like system, that's how it schedules different goroutines. It also supports CGo, which means you can call C functions (or Forth functions that adhere to the same ABI) from Go. It does not inherently need a main loop though that's how TinyGo programs are often structured.
For Forth, I'd recommend starting a new goroutine in the main function and then calling out to Forth. TinyGo also supports exporting functions to C, that way you can call `time.Sleep` etc to put the Forth goroutine to sleep for a bit.

Unfortunately https://www.spyr.ch appears to be down at the moment so I can't open the links you gave.
